### ISSUE TITLE
WSAPI API (createHostSet) fails with a certain character in a comment

### ISSUE THREAD

**Issue by tai** [hpe-storage/python-3parclient#16]
I'm trying to embed JSON-encoded string (for operational reasons) in a comment field, and API call is failing with a HTTP 500 error.

Here is a CLI example which passes without an error:

    cli% createhostset -comment {["some", "json", "data"]} fooset
    cli% showhostset -d fooset
    Id Name   Members Comment
    46 fooset --      ["some", "json", "data"]

At this point, I can successfully read above hostset info over WSAPI:

    In [28]: cl.getHostSets()
    Out[28]:
    {'members': [{'comment': '["some", "json", "data"]',
       'id': 47,
       'name': 'fooset'}],
     'total': 1}

However, when I try to do the same thing (= create a hostset over WSAPI), it fails.
It seems either "[" or "]" is causing this, but any printable chracters up to 255 characters should be supported as stated in the documentation (CLI reference, createhostset command).

    In [32]: data = '["some", "json", "data"]'
    In [33]: cl.createHostSet("fooset", comment=data)
    ---------------------------------------------------------------------------
    HTTPInternalServerError                   Traceback (most recent call last)
    /d/src/gio-boot/gio/storage/hp3par.py in <module>()
    ----> 1 cl.createHostSet("fooset", comment=data)
   
    /opt/rh/python33/root/usr/lib/python3.3/site-packages/hp3parclient-3.2.2-py3.3.egg/hp3parclient/client.py in createHostSet(self, name, domain, comment, setmembers)
       1266             info = self._mergeDict(info, members)
       1267
    -> 1268         response, body = self.http.post('/hostsets', body=info)
       1269         if response is not None and 'location' in response:
       1270             host_set_id = response['location'].rsplit(
    
    /opt/rh/python33/root/usr/lib/python3.3/site-packages/hp3parclient-3.2.2-py3.3.egg/hp3parclient/http.py in post(self, url, **kwargs)
        287                          will be an object
        288         """
    --> 289         return self._cs_request(url, 'POST', **kwargs)
        290
        291     def put(self, url, **kwargs):
    
    /opt/rh/python33/root/usr/lib/python3.3/site-packages/hp3parclient-3.2.2-py3.3.egg/hp3parclient/http.py in _cs_request(self, url, method, **kwargs)
        232         try:
        233             resp, body = self._time_request(self.api_url + url, method,
    --> 234                                             **kwargs)
        235             return resp, body
        236         except exceptions.HTTPUnauthorized as ex:
    
    /opt/rh/python33/root/usr/lib/python3.3/site-packages/hp3parclient-3.2.2-py3.3.egg/hp3parclient/http.py in _time_request(self, url, method, **kwargs)
        208     def _time_request(self, url, method, **kwargs):
        209         start_time = time.time()
    --> 210         resp, body = self.request(url, method, **kwargs)
        211         self.times.append(("%s %s" % (method, url),
        212                            start_time, time.time()))
    
    /opt/rh/python33/root/usr/lib/python3.3/site-packages/hp3parclient-3.2.2-py3.3.egg/hp3parclient/http.py in request(self, *args, **kwargs)
        202
        203         if resp.status >= 400:
    --> 204             raise exceptions.from_response(resp, body)
        205
        206         return resp, body
    
    HTTPInternalServerError: Internal Server Error (HTTP 500) 1 - internal server error

As the error is HTTP 500, could this be an issue with the WSAPI server running on a 3PAR box?

**Comment by rasselin** [comment]
HI Tai,

Enable debug logging. This will show you the full HTTP requests and responses  between the client and the server. 

API call: https://github.com/hp-storage/python-3parclient/blob/master/hp3parclient/client.py#L246
Example: https://github.com/hp-storage/python-3parclient/blob/master/samples/test_client.py#L23

Ramy

**Comment by tai** [comment]
Sure. Here it it:

```
In [9]: data = '["some", "json", "data"]'
In [10]: cl.createHostSet("fooset", comment=data)
REQ: curl -i http://10.57.14.138:8008/api/v1/hostsets -X POST -H "X-Hp3Par-Wsapi -Sessionkey: 0-c2e6cd07ac4384ed34ac0cb3f832110b-8978c455" -H "Content-Type: application/json" -H "Accept: application/json" -H "User-Agent: python-3parclient"

DEBUG:hp3parclient.http:
REQ: curl -i http://10.57.14.138:8008/api/v1/hostsets -X POST -H "X-Hp3Par-Wsapi -Sessionkey: 0-c2e6cd07ac4384ed34ac0cb3f832110b-8978c455" -H "Content-Type: application/json" -H "Accept: application/json" -H "User-Agent: python-3parclient"

REQ BODY: {"name": "fooset", "comment": "[\"some\", \"json\", \"data\"]"}

DEBUG:hp3parclient.http:REQ BODY: {"name": "fooset", "comment": "[\"some\", \"json\", \"data\"]"}

RESP:{'connection': 'close',
 'content-type': 'application/json',
 'date': 'Fri, 07 Aug 2015 09:22:30 GMT',
 'server': 'hp3par-wsapi',
 'status': '500'}

DEBUG:hp3parclient.http:RESP:{'connection': 'close',
 'content-type': 'application/json',
 'date': 'Fri, 07 Aug 2015 09:22:30 GMT',
 'server': 'hp3par-wsapi',
 'status': '500'}

RESP BODY:{"code":1,"desc":"internal server error"}
DEBUG:hp3parclient.http:RESP BODY:{"code":1,"desc":"internal server error"}

---------------------------------------------------------------------------
HTTPInternalServerError                   Traceback (most recent call last)
/d/src/gio-boot/gio/storage/hp3par.py in <module>()
----> 1 cl.createHostSet("fooset", comment=data)

/opt/rh/python33/root/usr/lib/python3.3/site-packages/hp3parclient-3.2.2-py3.3.egg/hp3parclient/client.py in createHostSet(self, name, domain, comment, setmembers)
   1266             info = self._mergeDict(info, members)
   1267
-> 1268         response, body = self.http.post('/hostsets', body=info)
   1269         if response is not None and 'location' in response:
   1270             host_set_id = response['location'].rsplit(

/opt/rh/python33/root/usr/lib/python3.3/site-packages/hp3parclient-3.2.2-py3.3.egg/hp3parclient/http.py in post(self, url, **kwargs)
    287                          will be an object
    288         """
--> 289         return self._cs_request(url, 'POST', **kwargs)
    290
    291     def put(self, url, **kwargs):

/opt/rh/python33/root/usr/lib/python3.3/site-packages/hp3parclient-3.2.2-py3.3.egg/hp3parclient/http.py in _cs_request(self, url, method, **kwargs)
    232         try:
    233             resp, body = self._time_request(self.api_url + url, method,
--> 234                                             **kwargs)
    235             return resp, body
    236         except exceptions.HTTPUnauthorized as ex:

/opt/rh/python33/root/usr/lib/python3.3/site-packages/hp3parclient-3.2.2-py3.3.egg/hp3parclient/http.py in _time_request(self, url, method, **kwargs)
    208     def _time_request(self, url, method, **kwargs):
    209         start_time = time.time()
--> 210         resp, body = self.request(url, method, **kwargs)
    211         self.times.append(("%s %s" % (method, url),
    212                            start_time, time.time()))

/opt/rh/python33/root/usr/lib/python3.3/site-packages/hp3parclient-3.2.2-py3.3.egg/hp3parclient/http.py in request(self, *args, **kwargs)
    202
    203         if resp.status >= 400:
--> 204             raise exceptions.from_response(resp, body)
    205
    206         return resp, body

HTTPInternalServerError: Internal Server Error (HTTP 500) 1 - internal server error
In [11]:
```
And FYI, here is what's really going over the wire:

```
POST /api/v1/hostsets HTTP/1.1
Host: 10.57.14.138:8008
Content-Length: 80
x-hp3par-wsapi-sessionkey: 0-c2e6cd07ac4384ed34ac0cb3f832110b-8978c455
content-type: application/json
accept: application/json
accept-encoding: gzip, deflate
user-agent: python-3parclient

{"name": "fooset", "comment": "{\"comment\": \"[\\\"some\\\", \\\"json\\\"]\"}"}
```

This nested JSON itself is valid, but not sure why "comment" field key is nested.
Could this be the cause of this error?

**Comment by kumartin** [comment]
Fixed in 3PAR firmware version 3.2.2 MU2
